### PR TITLE
[WIP] Node ipfs directories

### PR DESCRIPTION
**Pull request by DieMyst**

### CODE REVIEW

File: effects/ipfs/src/main/scala/fluence/effects/ipfs/IpfsStore.scala [hunk 1]
```diff
@@ -59,28 +66,84 @@ class IpfsStore[F[_]: Concurrent: ContextShift](ipfsUri: Uri)(
...
-          val errorMessage = s"IPFS download error $uri: $er"
-          IpfsError(errorMessage)
-        }
-        .map { r =>
-          logger.debug(s"IPFS download finished $uri")
-          r
-        }
-        .leftMap(identity[StoreError])
+    val uri = CatUri.param("arg", address)
+    logger.debug(s"IPFS download started $uri")
```
Comment (folex) [inline effects/ipfs/src/main/scala/fluence/effects/ipfs/IpfsStore.scala]: Let's wrap logger info F, so it is repeated via `backoff`

File: effects/ipfs/src/main/scala/fluence/effects/ipfs/IpfsStore.scala [hunk 2]
```diff
@@ -59,28 +66,84 @@ class IpfsStore[F[_]: Concurrent: ContextShift](ipfsUri: Uri)(
...
+        logger.debug(s"IPFS download finished $uri")
+        r
+      }
+      .leftMap(identity[StoreError])
+  }
+
+  private def lsRaw(hash: ByteVector): EitherT[F, StoreError, IpfsLsResponse] = {
+    val address = toAddress(hash)
+    val uri = LsUri.param("arg", address)
+    logger.debug(s"IPFS `ls` started $uri")
```
Comment (folex) [inline effects/ipfs/src/main/scala/fluence/effects/ipfs/IpfsStore.scala]: wrap log into F

File: effects/ipfs/src/main/scala/fluence/effects/ipfs/IpfsStore.scala [hunk 3]
```diff
@@ -59,28 +66,84 @@ class IpfsStore[F[_]: Concurrent: ContextShift](ipfsUri: Uri)(
...
+   * If hash belongs to file, returns the same hash.
+   *
+   * @param hash Content's hash
+   */
+  override def ls(hash: ByteVector): EitherT[F, StoreError, List[ByteVector]] =
+    for {
+      rawResponse <- lsRaw(hash)
+      _ <- assert(
+        rawResponse.Objects.size == 1,
+        IpfsError(s"One Object should be in IPFS response. Response: $rawResponse")
```
Comment (folex) [inline effects/ipfs/src/main/scala/fluence/effects/ipfs/IpfsStore.scala]: `Expected a single object, got ${rawResponse.Objects.size}. Response: $rawResponse`

File: effects/ipfs/src/main/scala/fluence/effects/ipfs/IpfsStore.scala [hunk 4]
```diff
@@ -59,28 +66,84 @@ class IpfsStore[F[_]: Concurrent: ContextShift](ipfsUri: Uri)(
...
+      )
+      rawHashes = {
+        val headObject = rawResponse.Objects.head
+        if (headObject.Links.isEmpty) List(headObject.Hash)
+        else headObject.Links.map(_.Hash)
+      }
+      hashes <- rawHashes.map { h =>
+        EitherT
+          .fromEither[F](fromAddress(h))
+          .leftMap(err => IpfsError(s"Cannot parse '$h' hex: $err").asInstanceOf[StoreError])
```
Comment (folex) [inline effects/ipfs/src/main/scala/fluence/effects/ipfs/IpfsStore.scala]: wouldn't safe type casting work here? i.e., `: StoreError`

Comment (folex) [reply effects/ipfs/src/main/scala/fluence/effects/ipfs/IpfsStore.scala]: You can also try with `flatMap(identity[StoreError])`

File: effects/ipfs/src/main/scala/fluence/effects/ipfs/IpfsStore.scala [hunk 5]
```diff
@@ -59,28 +66,84 @@ class IpfsStore[F[_]: Concurrent: ContextShift](ipfsUri: Uri)(
...
+      }.sequence
+    } yield {
+      logger.debug("List of file hashes: " + hashes)
```
Comment (folex) [inline effects/ipfs/src/main/scala/fluence/effects/ipfs/IpfsStore.scala]: Add prefix, e.g., `IPFS 'ls' hashes: ${hashes.mkString(" ")}`, and move that to trace

File: project/SbtCommons.scala [hunk 6]
```diff
@@ -100,8 +100,8 @@ object SbtCommons {
 
   val sttpVersion = "1.5.11"
   val sttp = "com.softwaremill.sttp"            %% "core"                           % sttpVersion
+  val sttpCats = "com.softwaremill.sttp"        %% "async-http-client-backend-cats" % "1.5.16-fl"
```
Comment (alari) [inline project/SbtCommons.scala]: Why `backend` suffix is absent?

File: effects/ipfs/src/main/scala/fluence/effects/ipfs/IpfsStore.scala (hunk 4 above)
Comment (alari) [reply effects/ipfs/src/main/scala/fluence/effects/ipfs/IpfsStore.scala]: Or `.leftMap[StoreError]`

**Pull request merged**